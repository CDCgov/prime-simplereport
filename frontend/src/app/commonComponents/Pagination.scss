@use "../../scss/settings" as *;

.usa-pagination {
  text-align: right;
  margin-right: 1rem;

  ol {
    list-style-type: none;
    padding-left: 2rem;
  }

  li {
    display: inline;
  }

  a,
  a:visited {
    padding: 4px 10px;
    text-decoration: none;
    color: $blue-cool-50;

    &.is-active {
      background-color: $blue-cool-70;
      border-radius: 100%;
      color: white;
    }
  }

  .fa-angle-right {
    margin-left: 3px;
  }
}
