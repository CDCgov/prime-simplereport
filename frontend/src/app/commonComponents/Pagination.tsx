import React from "react";
import classnames from "classnames";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faAngleLeft, faAngleRight } from "@fortawesome/free-solid-svg-icons";
import { IconProp } from "@fortawesome/fontawesome-svg-core";

import "./Pagination.scss";
import { LinkWithQuery } from "../commonComponents/LinkWithQuery";

interface Props {
  baseRoute: string;
  totalEntries: number;
  entriesPerPage: number;
  currentPage: number;
  pageGroupSize?: number;
  className?: string;
}

// Rules:
// * Always show first/last page (may be the same page!)
// * Show current page and N pages on either side
// * Use elipsis if gaps between group and first/last
//

// Always make this odd, so current page is in the middle
const defaultGroupSize = 7;
type PageListType = number | "firstEllipses" | "lastEllipses";
const Pagination = ({
  baseRoute,
  currentPage: rawCurrentPage,
  entriesPerPage,
  totalEntries,
  pageGroupSize = defaultGroupSize,
  className,
}: Props) => {
  const totalPages = Math.ceil(totalEntries / entriesPerPage);
  const currentPage = Math.min(Math.max(+rawCurrentPage || 0, 1), totalPages);
  const groupGutter = Math.floor(pageGroupSize / 2);
  const pageList: PageListType[] = [];

  // Ensure that the first, last, and at least groupGutter pages show
  let minGroupPage = currentPage - groupGutter;
  let maxGroupPage = currentPage + groupGutter;
  if (minGroupPage < 1) {
    maxGroupPage = maxGroupPage - minGroupPage + 1;
    minGroupPage = 1;
  }
  if (maxGroupPage > totalPages) {
    minGroupPage = Math.max(1, minGroupPage - (maxGroupPage - totalPages));
    maxGroupPage = totalPages;
  }

  const Link = (props: {
    to: string | number;
    active?: boolean;
    label?: string;
    children: React.ReactNode;
  }) => (
    <LinkWithQuery
      to={`${baseRoute}/${props.to}`}
      className={classnames(props.active && "is-active")}
      aria-label={props.label}
    >
      {props.children}
    </LinkWithQuery>
  );

  // Build list of pages, with 0 representing the ellipsis
  if (minGroupPage !== 1) {
    pageList.push(1);
  }
  if (minGroupPage > 2) {
    pageList.push("firstEllipses");
  }
  for (let pn = minGroupPage; pn <= maxGroupPage; pn++) {
    pageList.push(pn);
  }
  if (maxGroupPage < totalPages - 1) {
    pageList.push("lastEllipses");
  }
  if (maxGroupPage !== totalPages) {
    pageList.push(totalPages);
  }

  return (
    <nav
      className={classnames("usa-pagination", className)}
      role="navigation"
      aria-label="Pagination"
    >
      <ol>
        {currentPage > 1 && (
          <li key="prevpage">
            <Link to={`${currentPage - 1}`} label="Previous Page">
              <FontAwesomeIcon icon={faAngleLeft as IconProp} /> Prev
            </Link>
          </li>
        )}
        {pageList.map((pn) =>
          typeof pn === "number" ? (
            <li key={pn}>
              <Link to={pn} label={`Page ${pn}`} active={pn === currentPage}>
                <span>{pn}</span>
              </Link>
            </li>
          ) : (
            <li key={pn} aria-hidden="true">
              â€¦
            </li>
          )
        )}
        {currentPage < totalPages && (
          <li key="nextpage">
            <Link to={`${currentPage + 1}`} label="Next Page">
              Next <FontAwesomeIcon icon={faAngleRight as IconProp} />
            </Link>
          </li>
        )}
      </ol>
    </nav>
  );
};

export default Pagination;
