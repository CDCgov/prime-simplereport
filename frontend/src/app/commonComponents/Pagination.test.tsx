import { act, render, screen } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";

import Pagination from "./Pagination";

/*
// Cases (_n_ represents currentPage):
//  << _1_ >>
//  << 1 _2_ 3 >>
//  << _1_ 2 3 4 5 ... 71  >>
//  << 1 2 _3_ 4 5 ... 71  >>
//  << 1 ... 14 15 _16_ 17 18 ... 71 >>
//  << 1 ... 67 68 _69_ 70 71 >>
//  << 1 ... 67 68 69 70 _71_ >>
//  << 1 2 3 _4_ 5 6 ... 71 >>
*/

const defaults = {
  currentPage: 1,
  entriesPerPage: 10,
  totalEntries: 2,
  pageGroupSize: 5,
  className: undefined,
  baseRoute: "",
};

const testCases = {
  "with baseRoute": { baseRoute: "/patients" },
  "with className": { className: "unique-class-name" },
  "undefined currentPage": { currentPage: undefined },
  "2 pages": { totalEntries: 18, currentPage: 2 },
  "7 pages": { totalEntries: 61, currentPage: 4 },
  "10 pages": { entriesPerPage: 5, totalEntries: 48, currentPage: 8 },
};

describe("Pagination", () => {
  it("should render Pagination cases", () => {
    Object.entries(testCases).forEach(([name, tc], _index) => {
      const props = { ...defaults, ...tc } as any;
      const { container } = render(
        <MemoryRouter>
          <Pagination {...props} />
        </MemoryRouter>
      );
      expect(container).toMatchSnapshot(`Pagination '${name}'`);
    });
  });

  it("should handle onClick event", async () => {
    let onPaginationClick: jest.Mock = jest.fn();
    let mockProp = {
      ...defaults,
      onPaginationClick,
    };
    const props = { ...mockProp, ...testCases["with baseRoute"] } as any;
    render(
      <MemoryRouter>
        <Pagination {...props} />
      </MemoryRouter>
    );
    await act(async () => screen.getAllByText("1")[0].closest("a")?.click());
    expect(onPaginationClick).toHaveBeenCalledWith(1);
  });
});
