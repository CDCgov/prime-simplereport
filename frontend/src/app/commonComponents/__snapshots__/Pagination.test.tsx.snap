// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination should render Pagination cases: Pagination '2 pages' 1`] = `
<div>
  <nav
    aria-label="Pagination"
    class="usa-pagination"
    role="navigation"
  >
    <ol>
      <li>
        <a
          aria-label="Previous Page"
          class=""
          href="/1"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-angle-left "
            data-icon="angle-left"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 320 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M41.4 233.4c-12.5 12.5-12.5 32.8 0 45.3l160 160c12.5 12.5 32.8 12.5 45.3 0s12.5-32.8 0-45.3L109.3 256 246.6 118.6c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0l-160 160z"
              fill="currentColor"
            />
          </svg>
           Prev
        </a>
      </li>
      <li>
        <a
          aria-label="Page 1"
          class=""
          href="/1"
        >
          <span>
            1
          </span>
        </a>
      </li>
      <li>
        <a
          aria-label="Page 2"
          class="is-active"
          href="/2"
        >
          <span>
            2
          </span>
        </a>
      </li>
    </ol>
  </nav>
</div>
`;

exports[`Pagination should render Pagination cases: Pagination '7 pages' 1`] = `
<div>
  <nav
    aria-label="Pagination"
    class="usa-pagination"
    role="navigation"
  >
    <ol>
      <li>
        <a
          aria-label="Previous Page"
          class=""
          href="/3"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-angle-left "
            data-icon="angle-left"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 320 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M41.4 233.4c-12.5 12.5-12.5 32.8 0 45.3l160 160c12.5 12.5 32.8 12.5 45.3 0s12.5-32.8 0-45.3L109.3 256 246.6 118.6c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0l-160 160z"
              fill="currentColor"
            />
          </svg>
           Prev
        </a>
      </li>
      <li>
        <a
          aria-label="Page 1"
          class=""
          href="/1"
        >
          <span>
            1
          </span>
        </a>
      </li>
      <li>
        <a
          aria-label="Page 2"
          class=""
          href="/2"
        >
          <span>
            2
          </span>
        </a>
      </li>
      <li>
        <a
          aria-label="Page 3"
          class=""
          href="/3"
        >
          <span>
            3
          </span>
        </a>
      </li>
      <li>
        <a
          aria-label="Page 4"
          class="is-active"
          href="/4"
        >
          <span>
            4
          </span>
        </a>
      </li>
      <li>
        <a
          aria-label="Page 5"
          class=""
          href="/5"
        >
          <span>
            5
          </span>
        </a>
      </li>
      <li>
        <a
          aria-label="Page 6"
          class=""
          href="/6"
        >
          <span>
            6
          </span>
        </a>
      </li>
      <li>
        <a
          aria-label="Page 7"
          class=""
          href="/7"
        >
          <span>
            7
          </span>
        </a>
      </li>
      <li>
        <a
          aria-label="Next Page"
          class=""
          href="/5"
        >
          Next 
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-angle-right "
            data-icon="angle-right"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 320 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M278.6 233.4c12.5 12.5 12.5 32.8 0 45.3l-160 160c-12.5 12.5-32.8 12.5-45.3 0s-12.5-32.8 0-45.3L210.7 256 73.4 118.6c-12.5-12.5-12.5-32.8 0-45.3s32.8-12.5 45.3 0l160 160z"
              fill="currentColor"
            />
          </svg>
        </a>
      </li>
    </ol>
  </nav>
</div>
`;

exports[`Pagination should render Pagination cases: Pagination '10 pages' 1`] = `
<div>
  <nav
    aria-label="Pagination"
    class="usa-pagination"
    role="navigation"
  >
    <ol>
      <li>
        <a
          aria-label="Previous Page"
          class=""
          href="/7"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-angle-left "
            data-icon="angle-left"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 320 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M41.4 233.4c-12.5 12.5-12.5 32.8 0 45.3l160 160c12.5 12.5 32.8 12.5 45.3 0s12.5-32.8 0-45.3L109.3 256 246.6 118.6c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0l-160 160z"
              fill="currentColor"
            />
          </svg>
           Prev
        </a>
      </li>
      <li>
        <a
          aria-label="Page 1"
          class=""
          href="/1"
        >
          <span>
            1
          </span>
        </a>
      </li>
      <li
        aria-hidden="true"
      >
        â€¦
      </li>
      <li>
        <a
          aria-label="Page 6"
          class=""
          href="/6"
        >
          <span>
            6
          </span>
        </a>
      </li>
      <li>
        <a
          aria-label="Page 7"
          class=""
          href="/7"
        >
          <span>
            7
          </span>
        </a>
      </li>
      <li>
        <a
          aria-label="Page 8"
          class="is-active"
          href="/8"
        >
          <span>
            8
          </span>
        </a>
      </li>
      <li>
        <a
          aria-label="Page 9"
          class=""
          href="/9"
        >
          <span>
            9
          </span>
        </a>
      </li>
      <li>
        <a
          aria-label="Page 10"
          class=""
          href="/10"
        >
          <span>
            10
          </span>
        </a>
      </li>
      <li>
        <a
          aria-label="Next Page"
          class=""
          href="/9"
        >
          Next 
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-angle-right "
            data-icon="angle-right"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 320 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M278.6 233.4c12.5 12.5 12.5 32.8 0 45.3l-160 160c-12.5 12.5-32.8 12.5-45.3 0s-12.5-32.8 0-45.3L210.7 256 73.4 118.6c-12.5-12.5-12.5-32.8 0-45.3s32.8-12.5 45.3 0l160 160z"
              fill="currentColor"
            />
          </svg>
        </a>
      </li>
    </ol>
  </nav>
</div>
`;

exports[`Pagination should render Pagination cases: Pagination 'undefined currentPage' 1`] = `
<div>
  <nav
    aria-label="Pagination"
    class="usa-pagination"
    role="navigation"
  >
    <ol>
      <li>
        <a
          aria-label="Page 1"
          class="is-active"
          href="/1"
        >
          <span>
            1
          </span>
        </a>
      </li>
    </ol>
  </nav>
</div>
`;

exports[`Pagination should render Pagination cases: Pagination 'with baseRoute' 1`] = `
<div>
  <nav
    aria-label="Pagination"
    class="usa-pagination"
    role="navigation"
  >
    <ol>
      <li>
        <a
          aria-label="Page 1"
          class="is-active"
          href="/patients/1"
        >
          <span>
            1
          </span>
        </a>
      </li>
    </ol>
  </nav>
</div>
`;

exports[`Pagination should render Pagination cases: Pagination 'with className' 1`] = `
<div>
  <nav
    aria-label="Pagination"
    class="usa-pagination unique-class-name"
    role="navigation"
  >
    <ol>
      <li>
        <a
          aria-label="Page 1"
          class="is-active"
          href="/1"
        >
          <span>
            1
          </span>
        </a>
      </li>
    </ol>
  </nav>
</div>
`;
